diff --git a/src/Composer/Installers/DockerInstaller.php b/src/Composer/Installers/DockerInstaller.php
index e69de29..3b4ebfd 100644
--- a/src/Composer/Installers/DockerInstaller.php
+++ b/src/Composer/Installers/DockerInstaller.php
@@ -0,0 +1,9 @@
+<?php
+namespace Composer\Installers;
+
+class DockerInstaller extends BaseInstaller
+{
+    protected $locations = array(
+        'compose' => '{$name}/',
+    );
+}
\ No newline at end of file
diff --git a/src/Composer/Installers/Installer.php b/src/Composer/Installers/Installer.php
index 352cb7f..804ffe9 100644
--- a/src/Composer/Installers/Installer.php
+++ b/src/Composer/Installers/Installer.php
@@ -38,6 +38,7 @@ class Installer extends LibraryInstaller
         'dokuwiki'     => 'DokuWikiInstaller',
         'dolibarr'     => 'DolibarrInstaller',
         'decibel'      => 'DecibelInstaller',
+        'docker'       => 'DockerInstaller',
         'drupal'       => 'DrupalInstaller',
         'elgg'         => 'ElggInstaller',
         'eliasis'      => 'EliasisInstaller',
